src/vs/workbench/test/electron-sandbox/resolveExternal.test.ts

```diff
@@ -0,0 +1,123 @@
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+import * as assert from 'assert';
+import { ensureNoDisposablesAreLeakedInTestSuite } from 'vs/base/test/common/utils';
+import { NativeWindow } from 'vs/workbench/electron-sandbox/window';
+import { ITunnelService, RemoteTunnel } from 'vs/platform/tunnel/common/tunnel';
+import { URI } from 'vs/base/common/uri';
+import { TestInstantiationService } from 'vs/platform/instantiation/test/common/instantiationServiceMock';
+import { IAddressProvider } from 'vs/platform/remote/common/remoteAgentConnection';
+import { workbenchInstantiationService } from 'vs/workbench/test/electron-sandbox/workbenchTestServices';
+import { DisposableStore } from 'vs/base/common/lifecycle';
+
+type PortMap = Record<number, number>;
+
+class TunnelMock implements Partial<ITunnelService> {
+	private assignedPorts: PortMap = {};
+	private expectedDispose = false;
+
+	reset(ports: PortMap) {
+		this.assignedPorts = ports;
+	}
+
+	expectDispose() {
+		this.expectedDispose = true;
+	}
+
+	getExistingTunnel(): Promise<string | RemoteTunnel | undefined> {
+		return Promise.resolve(undefined);
+	}
+
+	openTunnel(_addressProvider: IAddressProvider | undefined, _host: string | undefined, port: number): Promise<RemoteTunnel | string | undefined> | undefined {
+		if (!this.assignedPorts[port]) {
+			return Promise.reject(new Error('Unexpected tunnel request'));
+		}
+		const res: RemoteTunnel = {
+			localAddress: `localhost:${this.assignedPorts[port]}`,
+			tunnelRemoteHost: '4.3.2.1',
+			tunnelRemotePort: this.assignedPorts[port],
+			privacy: '',
+			dispose: () => {
+				assert(this.expectedDispose, 'Unexpected dispose');
+				this.expectedDispose = false;
+				return Promise.resolve();
+			}
+		};
+		delete this.assignedPorts[port];
+		return Promise.resolve(res);
+	}
+
+	validate() {
+		try {
+			assert(Object.keys(this.assignedPorts).length === 0, 'Expected tunnel to be used');
+			assert(!this.expectedDispose, 'Expected dispose to be called');
+		} finally {
+			this.expectedDispose = false;
+		}
+	}
+}
+
+class TestNativeWindow extends NativeWindow {
+	protected override create(): void { }
+	protected override registerListeners(): void { }
+	protected override enableMultiWindowAwareTimeout(): void { }
+}
+
+suite('NativeWindow:resolveExternal', () => {
+	const disposables = new DisposableStore();
+	const tunnelMock = new TunnelMock();
+	let window: TestNativeWindow;
+
+	async function doTest(uri: string, ports: PortMap = {}, expectedUri?: string) {
+		const instantiationService: TestInstantiationService = <TestInstantiationService>workbenchInstantiationService(undefined, disposables);
+		instantiationService.stub(ITunnelService, tunnelMock);
+		window = disposables.add(instantiationService.createInstance(TestNativeWindow));
```

Comment:
Is there a way to instantiate a single instance for all the test suite, and dispose it after all tests are done? This is a lot of work, while the function I run from this mock is stateless...

Reply:
Yes, `suiteSetup` and `suiteTeardown`. You can see an example here: 

https://github.com/microsoft/vscode/blob/c37ca48a96e37e30e13ed4d4679bca1c2e330ad4/src/vs/workbench/api/test/common/extensionHostMain.test.ts#L75-L82

Reply:
Done